chaincode/src/burns/burnTokens.ts

```diff
@@ -209,3 +171,167 @@ export async function burnTokens(
+ * at most one value per unique `tokenInstanceKey`.
+ *
+ * @param requests: BurnTokenQuantity[]
+ * @returns BurnTokenQuantity[]
+ */
+export function aggregateBurnQuantities(requests: BurnTokenQuantity[]): BurnTokenQuantity[] {
+  const summedQuantities: BurnTokenQuantity[] = [];
+  const hash = {};
+
+  for (let i = 0; i < requests.length; i++) {
+    const { collection, category, type, additionalKey, instance } = requests[i].tokenInstanceKey;
+
+    const key = `${collection}_${category}_${type}_${additionalKey}_${instance.toString()}`;
+
+    hash[key] = hash[key] ?? summedQuantities.length;
+
+    const tokenIndex = hash[key];
+
+    if (summedQuantities[tokenIndex] === undefined) {
+      summedQuantities[tokenIndex] = instanceToInstance(requests[i]);
+    } else {
+      summedQuantities[tokenIndex].quantity = summedQuantities[tokenIndex].quantity.plus(
+        requests[i].quantity
+      );
+    }
+  }
+
+  return summedQuantities;
+}
+
+/**
+ * If the `TokenBurn` can be read, increment the quantity.
+ * Otherwise, create a new `TokenBurn` with the quantity set to the provided value.
+ *
+ * @remarks
+ * The ChainKey design of `TokenBurn` objects ensures uniqueness by
+ * burnedBy, tokenInstanceKey properties, and a created timestamp.
+ * Generally if a `TokenBurn` is read twice in the same transaction it means
+ * two `burnQuantities` were provided to `burnTokens` for the same token instance,
+ * or two separate functions burned the same token (e.g. a fee and a chaincode main method execution.)
+ */
+export async function incrementOrCreateTokenBurnForTx(
+  ctx: GalaChainContext,
+  burnedBy: string,
+  { collection, category, type, additionalKey, instance }: TokenInstanceKey,
+  quantity: BigNumber
+): Promise<TokenBurn> {
+  const instanceKey: TokenInstanceKey = plainToInstance(TokenInstanceKey, {
+    collection,
+    category,
+    type,
+    additionalKey,
+    instance
+  });
+
+  const newBurn = await createValidChainObject(TokenBurn, {
+    burnedBy,
+    collection,
+    category,
+    type,
+    additionalKey,
+    instance,
+    created: ctx.txUnixTime,
+    quantity
+  });
+
+  const burnKey = newBurn.getCompositeKey();
+
+  const cachedBurn: TokenBurn | undefined = await getObjectByKey(ctx, TokenBurn, burnKey).catch((e) => {
+    const chainError = ChainError.from(e);
+    if (chainError.matches(ErrorCode.NOT_FOUND)) {
+      return undefined;
+    } else {
+      throw chainError;
+    }
+  });
+
+  if (cachedBurn !== undefined) {
+    cachedBurn.quantity = cachedBurn.quantity.plus(quantity);
+
+    return cachedBurn;
+  } else {
+    return newBurn;
+  }
+}
+
+/**
+ * If the `TokenBurnCounter` can be read, increment the quantity.
+ * Otherwise, create a new `TokenBurnCounter` with the quantity set to the provided value.
+ *
+ * @remarks
+ * The ChainKey design of `TokenBurnCounter` ranged objects ensures uniqueness by
+ * tokenInstanceKey properties, burnedBy, and a timeKey derived from the transaction timestamp.
```

Comment:
technically, the transaction timestamp resolution is in seconds, which makes it unreliable if we want to use it for ensuring uniqueness. Don't you think it would be safer to use transaction ID instead? Is it a big deal?

Reply:
I agree its worth revisiting, however I did not want to increase the scope of this fix so I left as much as-is as I could. 

I imagine if we were to rework some chain keys we might also end up with a  data migration of some sort on our hands? And/or what downstream implications that could have on data consumers. 

Also, I think I looked into this before, and found that the `ctx.txUnixTime` is in fact giving us a timestamp in milliseconds. I could be mis-remembering. But if `fabric-node-sdk` is using JavaScript's `Date.now()` under the hood, it could explain why it returns a milliseconds timestamp. 

Milliseconds certainly add a bit more specificity, but I still agree that using a timestamp as the most specific chain key on any chain object (like `TokenBurn`) is always ill-advised and the transaction id is a better key for uniqueness.